### PR TITLE
Fixed #13312 --  Added a way to customize the order_by() of null fields.

### PR DESCRIPTION
**Pull request by boblefrag**
https://code.djangoproject.com/ticket/13312

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -889,6 +893,30 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
...
+        if self.nullslast:
+            template = '%(expression)s %(ordering)s NULLS LAST'
+        sql, params = self.as_sql(compiler, connection, template=template)
+        return sql, params
+
+    def as_oracle(self, compiler, connection):
+        return self.as_postgresql(self, compiler, connection)
+
+    def as_mysql(self, compiler, connection):
+        template = None
```
Comment (charettes) [inline django/db/models/expressions.py]: Unused variable.

File: django/db/models/expressions.py [hunk 2]
```diff
@@ -889,6 +893,30 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
...
+        if self.nullslast:
+            template = 'IF(ISNULL(%(expression)s),1,0),%(expression)s %(ordering)s '
+            sql, params = self.as_sql(compiler, connection, template=template)
```
Comment (charettes) [inline django/db/models/expressions.py]: Is there a reason not to use the pattern used in other `as_*` methods?

Comment (boblefrag) [inline django/db/models/expressions.py]: No, my bad. Fixed

**Comment by charettes** [comment]
Buildbot, add to whitelist.

**Comment by charettes** [comment]
I suppose we could make `OrderBy.asc()/desc()` chainable in another commit by making them return `self`.

**Comment by charettes** [comment]
The tests are failing of PostgreSQL. Could you investigate?

**Comment by boblefrag** [comment]
Sure. I'm on it

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -889,6 +893,30 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
...
         return (template % placeholders).rstrip(), params
 
+    def as_sqlite(self, compiler, connection):
+        template = None
+        if self.nullslast:
+            template = '%(expression)s IS NULL, %(expression)s %(ordering)s'
+        sql, params = self.as_sql(compiler, connection, template=template)
+        return sql, params
+
+    def as_postgresql(self, compiler, connection):
```
Comment (jarshwah) [inline django/db/models/expressions.py]: It'd be better to rename this `as_sql` and to remove the `as_oracle` method since it is the "default" implementation that all backends follow unless they opt out. As-is, any third party backends *must* implement their own version even if they support NULLS LAST syntax, such as mssql.

**Comment by boblefrag** [comment]
@charettes now asc and desc are chainable with nulls_last as asked.

File: django/db/models/expressions.py [hunk 4]
```diff
@@ -879,6 +883,8 @@ def get_source_expressions(self):
         return [self.expression]
 
     def as_sql(self, compiler, connection, template=None, **extra_context):
+        if not template and self.nullslast:
+            template = '%(expression)s %(ordering)s NULLS LAST'
```
Comment (charettes) [inline django/db/models/expressions.py]: Use `"%s NULLS LAST" % self.template` to prevent code duplication.

File: django/db/models/expressions.py [hunk 5]
```diff
@@ -889,6 +895,20 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
         template = template or self.template
         return (template % placeholders).rstrip(), params
 
+    def as_sqlite(self, compiler, connection):
+        template = None
+        if self.nullslast:
+            template = '%(expression)s IS NULL, %(expression)s %(ordering)s'
+        sql, params = self.as_sql(compiler, connection, template=template)
+        return sql, params
```
Comment (charettes) [inline django/db/models/expressions.py]: Return directly.

File: django/db/models/expressions.py [hunk 6]
```diff
@@ -889,6 +895,20 @@ def as_sql(self, compiler, connection, template=None, **extra_context):
...
+
+    def as_mysql(self, compiler, connection):
+        template = None
+        if self.nullslast:
+            template = 'IF(ISNULL(%(expression)s),1,0),%(expression)s %(ordering)s '
+        sql, params = self.as_sql(compiler, connection, template=template)
+        return sql, params
```
Comment (charettes) [inline django/db/models/expressions.py]: Ditto.

**Comment by charettes** [comment]
Should we also add support for `NULLS FIRST`?

**Comment by charettes** [comment]
I feel like making `nulls` a parameter of `asc()`/`desc()` would make a better API than chaining by exposing both a `nulls_first()` and `nulls_last()` method. e.g.:

```python
Event.objects.order_by(F('start_date').asc(nulls='first')) or
Event.objects.order_by(F('start_date').asc(nulls_first=True))
```

**Comment by boblefrag** [comment]
@charette now comply with the api you suggest.

**Comment by charettes** [comment]
@jarshwah, @mjtamlyn what you think of the `(nulls='first')` API? The only thing that bugs me is the usage of a string constant as it not used anywhere else in the ORM. We could also take a tri-valued `nulls_first` (or `nulls_last` kwarg) where `False` maps to `NULLS LAST` and `None` does nothing.

**Comment by jarshwah** [comment]
Don't like the string argument, sorry, and I feel that a tri-value would constantly require referencing the docs. I'd prefer `nulls_first=True` or `nulls_last=True` as an argument to asc/desc. Are there any better options?

**Comment by charettes** [comment]
> Are there any better options?

I can't think of any. Maybe the this should be taken to the developpers mailing list?

**Comment by boblefrag** [comment]
I am ok with nulls_first=True syntax but before implementing it I need a green flag of yours so just let me know the endorsed syntax and I'll implement it.

**Comment by mjtamlyn** [comment]
I like the `nulls_first=True` option best as well. It's not brilliant, but best of a bad bunhc.

**Comment by charettes** [comment]
@boblefrag let's go with `nulls_first=True` and `nulls_last=True`.

You might want to delegate `asc()`/`desc()`'s `**kwargs` to `OrderBy.__init__()` and raise a `ValueError` if both `nulls_first` and  `nulls_last` are specified.

**Comment by adamchainz** [comment]
FWIW `nulls_first=True` and `nulls_last=True` seems like a nice solution to me too (that said using a set of allowed strings like `nulls='first'` isn't exactly unpythonic per se, c.f. `str.encode('ascii', errors='replace')`).

**Comment by boblefrag** [comment]
@charettes nulls_first=True and nulls_last=True. What's next step?

**Comment by charettes** [comment]
The code is LGTM. Maybe break the `nulls_first` and `nulls_last` options assertions in two tests and add an additional test method for the `ValueError` raised when both options are used together.

**Comment by boblefrag** [comment]
@charettes done.

File: tests/ordering/tests.py [hunk 7]
```diff
@@ -90,6 +90,90 @@ def test_order_by_override(self):
             attrgetter("headline")
         )
 
+    def test_order_by_nulls(self):
+        with self.assertRaises(ValueError):
```
Comment (charettes) [inline tests/ordering/tests.py]: Use `assertRaisesMessage`.

File: tests/ordering/tests.py [hunk 8]
```diff
@@ -90,6 +90,90 @@ def test_order_by_override(self):
             attrgetter("headline")
         )
 
+    def test_order_by_nulls(self):
```
Comment (charettes) [inline tests/ordering/tests.py]: Use a more descriptive name. `test_order_by_nulls_first_and_last`?

File: tests/ordering/tests.py [hunk 9]
```diff
@@ -90,6 +90,90 @@ def test_order_by_override(self):
...
+        with self.assertRaises(ValueError):
+            Article.objects.order_by(F("author").desc(nulls_last=True, nulls_first=True))
+
+    def test_order_by_nulls_last(self):
+        author_1 = Author.objects.create()
+        Article.objects.create(
+            headline="Article 5",
+            author=author_1,
+            pub_date=datetime(2005, 7, 28)
```
Comment (charettes) [inline tests/ordering/tests.py]: Add a trailing comma.

File: tests/ordering/tests.py [hunk 10]
```diff
@@ -90,6 +90,90 @@ def test_order_by_override(self):
...
+        )
+
+        author_2 = Author.objects.create()
+        Article.objects.create(
+            headline="Article 6",
+            author=author_2,
+            pub_date=datetime(2005, 7, 28)
```
Comment (charettes) [inline tests/ordering/tests.py]: Add a trailing comma.

**Comment by charettes** [comment]
You can reuse the `Article` objects created in setup instead of creating two new ones per tests by adding an author to one of them. You could also convert the test class to use `setupTestData` instead of `setUp`. Please use the title of this PR as commit message.

**Comment by boblefrag** [comment]
@charettes is there anything left to be donefor making this PR mergeable ?

**Comment by charettes** [comment]
Marked the ticket as ready for final review.

**Comment by timgraham** [comment]
Documentation is missing. See our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) for details. Please uncheck "Needs documentation" on the ticket after updating.

File: django/db/models/expressions.py [hunk 11]
```diff
@@ -862,7 +862,12 @@ def as_sql(self, compiler, connection, template=None, case_joiner=None, **extra_
 class OrderBy(BaseExpression):
     template = '%(expression)s %(ordering)s'
 
-    def __init__(self, expression, descending=False):
+    def __init__(self, expression, descending=False, nulls_first=False,
+                 nulls_last=False):
+        if nulls_first and nulls_last:
+            raise ValueError("nulls_first and nulls_last are mutualy exclusive")
```
Comment (danielsamuels) [inline django/db/models/expressions.py]: Minor typo: should be 'mutually'.

File: tests/ordering/tests.py [hunk 12]
```diff
@@ -90,6 +96,58 @@ def test_order_by_override(self):
             attrgetter("headline")
         )
 
+    def test_order_by_nulls_first_and_last(self):
+        with self.assertRaisesMessage(ValueError,
+                                      "nulls_first and nulls_last are mutualy exclusive"):
```
Comment (danielsamuels) [inline tests/ordering/tests.py]: Same 'mutually' typo here.

**Comment by boblefrag** [comment]
fixed typo and added doc

File: tests/ordering/tests.py [hunk 13]
```diff
@@ -10,19 +10,25 @@
 
 
 class OrderingTests(TestCase):
-    def setUp(self):
-        self.a1 = Article.objects.create(
+
+    @classmethod
```
Comment (timgraham) [inline tests/ordering/tests.py]: The change is fine, I'm just asking that it be a be separate commit similar to 31098e3288595c13f165935f2579f1af744e5240 so that the refactoring isn't mixed in with the new feature.

**Comment by boblefrag** [comment]
@timgraham here is the 2 commits. Let me know if it's ok for you.

File: tests/ordering/tests.py [hunk 14]
```diff
@@ -90,6 +96,53 @@ def test_order_by_override(self):
...
+        self.assertSequenceEqual(
+            Article.objects.order_by(F("author").asc(nulls_last=True)), [
+                self.a3,
+                self.a4,
+                self.a1,
+                self.a2
+            ]
+        )
+
+    def test_order_by_nulls_first(self):
```
Comment (felixxm) [inline tests/ordering/tests.py]: Please add authors for `article3` and `article4` like in `test_order_by_nulls_last`:
```diff
--- a/tests/ordering/tests.py
+++ b/tests/ordering/tests.py
@@ -125,6 +125,8 @@ class OrderingTests(TestCase):
         )
 
     def test_order_by_nulls_first(self):
+        Article.objects.filter(headline="Article 3").update(author=self.author_1)
+        Article.objects.filter(headline="Article 4").update(author=self.author_2)
         self.assertSequenceEqual(
             Article.objects.order_by(F("author").asc(nulls_first=True)), [
                 self.a1,
@@ -138,8 +140,8 @@ class OrderingTests(TestCase):
             Article.objects.order_by(F("author").desc(nulls_first=True)), [
                 self.a1,
                 self.a2,
-                self.a3,
-                self.a4
+                self.a4,
+                self.a3
             ]
         )
```

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/7563.
